### PR TITLE
server: let select for update could be recorded for slow log and statements

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve?

Problem Summary:
The pessimistic `select for update` statements could not be logged as slow query, also other things `RecordSet`.`Close` will do

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

Change the close function of  `chunkRowRecordSet` 

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)
Start two sessions let the second select for update statement wait lock for sometime, the slow query should recorded like
```
# Conn_ID: 2
# Query_time: 8.580836624
# Parse_time: 0.000062642
# Compile_time: 0.000029461
# LockKeys_time: 8.580678652
# DB: test
# Is_internal: false
# Digest: 0cef963ebac4619e5a5c4562d83626a3abb5df2b0501a4dbeae157f7e5356cff
# Num_cop_tasks: 0
# Prepared: false
# Has_more_results: false
# Succ: true
# Plan: tidb_decode_plan('JIwwCTM2XzEJMAkxCXRhYmxlOnQxLCBoYW5kbGU6MiwgbG9jawo=')
# Plan_digest: 3f5625f5b5b7a079571f7844fe4e63a6587a0674dace590987cdb159574bdd22
select * from t1 where c1 = 2 for update;
```

Side effects

### CODE REVIEW

**Comment by cfzjywxk** [comment]
@jackysp @coocood @lysu @tiancaiamao  PTAL

**Comment by jackysp** [comment]
Was it introduced by https://github.com/pingcap/tidb/pull/15890 ?

File: executor/adapter.go [hunk 1]
```diff
@@ -437,6 +438,12 @@ func (c *chunkRowRecordSet) NewChunk() *chunk.Chunk {
 }
 
 func (c *chunkRowRecordSet) Close() error {
+	c.execStmt.FinishExecuteStmt(c.execStmt.Ctx.GetSessionVars().TxnCtx.StartTS, true, false)
```
Comment (coocood) [inline executor/adapter.go:441]: Can we extract the line to a method in `ExecStmt`?

**Comment by cfzjywxk** [comment]
> Was it introduced by #15890 ?

The original logic will log each pessimistic retry, it was fixed in #15890, but it does not get logged

**Comment by coocood** [comment]
LGTM

**Comment by cfzjywxk** [comment]
cherry pick to release-4.0 in PR #16897

**Comment by cfzjywxk** [comment]
cherry pick to release-3.1 in PR #16901

**Comment by cfzjywxk** [comment]
cherry pick to release-3.0 in PR #16903

**Pull request merged**